### ISSUE TITLE
How do 64-bit chugins on Windows work?

### ISSUE THREAD

**Issue by DBraun** [ccrma/chugins#45]
I realized that for the VST plugin I'm working on, I'll want to load 64-bit VSTs instead of 32-bit. Once I started compiling the chugin to 64-bit instead of 32-bit, my VST.chug is no longer recognized by ChucK. How do 64-bit chugins work on Windows? Is there hope for a 64-bit VST chugin on Windows?

the chuck console says `undefined type 'VST'...` when it previously didn't have a problem. By the way my preprocessor statements include 

```
        __WINDOWS_MODERN__
        __PLATFORM_WIN32__
        __WINDOWS_DS__
        _WINDOWS
        WIN32
        _USRDLL
        NDEBUG
```

chuck code:
```chuck
SndBuf b => VST vst => dac;
"special:dope" => b.read;

if (vst.loadplugin("C:/VSTPlugIns/64bitVST_x64.dll")) {
   <<< "success!" >>>;
} else {
    <<< "fail!" >>>;
}

while( true )
{
    0 => b.pos;
    200::ms => now;
}
```

Is this bug still happening? https://lists.cs.princeton.edu/pipermail/chuck-users/2018-May/008369.html
Should the ChucK installer have put ChucK in C:/Program Files/ChucK not C:/Program Files (x86)/ChucK?

I've tried placing the 64-bit VST.chug in C:/Program Files/ChucK/chugins and C:/Program Files (x86)/ChucK/chugins.

This is the chuck console output:

[chuck]:(3:SEVERE):  |  |  | loading chugin 'VST.chug'
[chuck]:(3:SEVERE):  |  |  | error loading chugin 'VST.chug', skipping
[chuck]:(3:SEVERE):  |  |  | error from chuck_dl: '193'

Does 193 mean "ERROR_BAD_EXE_FORMAT"? https://docs.microsoft.com/en-us/windows/win32/debug/system-error-codes--0-499- 
https://github.com/ccrma/chuck/blob/05913b160e3fe881e5fb79d21d1d59a9fd9d02be/src/core/chuck_dl.cpp#L1190

https://stackoverflow.com/questions/38579909/loadlibrary-fails-with-error-code-193

"Error 193 is ERROR_BAD_EXE_FORMAT, most likely meaning the file is corrupted, or it's a 32-bit DLL you're loading into a 64-bit process or vice versa. "

testing from command line:
```
chuck --version
chuck version: 1.4.0.1 (numchucks)
   microsoft win32 : 32-bit
   http://chuck.cs.princeton.edu/
   http://chuck.stanford.edu/
```

How do I get the 64-bit version installed? http://chuck.stanford.edu/release/ Update: I built chuck.exe from the 1.4.0.1 source and now chuck --version is
```
chuck version: 1.4.0.1 (numchucks)
   microsoft win32 : 64-bit
   http://chuck.cs.princeton.edu/
   http://chuck.stanford.edu/
```
Then I replaced the chuck/include in this repo with the newer 1.4.0.1 source and rebuilt my chugin as 64-bit. However, it's still not being loaded due to the 193 error.

**Comment by lathertonj** [comment]
Good to see that you figured out it's possible to build ChucK yourself in 64-bit. I think it should be possible to rebuild the Chugins in 64-bit yourself as well?

**Comment by DBraun** [comment]
![2020-11-30 12_39_25-VST - Microsoft Visual Studio](https://user-images.githubusercontent.com/2096055/100662015-2eedf100-3309-11eb-8e47-85758ffa81ae.png)

Yes I'm pretty sure I'm building a 64-bit chugin, but it's still showing the 193 error in the chuck console with what seems like a 64-bit chuck. All I did was compile the chuck.exe from source and replace the chuck.exe in C:/Program Files (x86)/Chuck/bin

**Comment by DBraun** [comment]
I'm still working on this. I opened a native tools cmd and used

`/dumpbin /headers vst.dll > my64bit_headers.txt`
`/dumpbin /exports vst.dll > my64bit_exports.txt`

(I rename vst.dll to vst.chug when I try to use it in chuck.)

I think the outputs confirm that the chugin is 64-bit. I also used [dependency walker](https://github.com/lucasg/Dependencies) to see if anything in the dependency chain is 32-bit. It seems that nothing is. I used task manager to make sure chuck.exe is 64-bit. I'm still getting this 193 error. Tough one...

[my64bit_headers.txt](https://github.com/ccrma/chugins/files/5759647/my64bit_headers.txt)
[my64bit_exports.txt](https://github.com/ccrma/chugins/files/5759648/my64bit_exports.txt)

**Comment by DBraun** [comment]
Closing because I found and resolved the issue.
* miniAudicle is 32-bit, so I think no 64-bit chugin will work.
* I was running chuck.exe from the command line, but it wasn't looking at my custom folder holding the chugin. I needed to set an environment variable `CHUCK_CHUGIN_PATH` equal to `C:\my\custom\path\to\chugins`. I assume miniAudicle does this each time at startup. Or I could have just put the chugin in the default location: `C:\Program Files (x86)\ChucK\chugins`. Now everything works!
